lean4-infoview/src/infoview/userWidget.tsx

```diff
@@ -13,27 +14,82 @@ import { GoalsLocation } from './goalLocation'
+ * of the hash stored in `Lean.Widget.Module.javascriptHash`.
+ *
+ * In the future,
+ * we may support importing widget modules by their fully qualified Lean name
+ * (e.g. `'widget_module:name,Lean.Meta.Tactic.TryThis.tryThisWidget'`),
+ * or some way to assign widget modules a more NPM-friendly name
+ * so that the usual URIs (e.g. `'@leanprover-community/pro-widgets'`) work.
+ */
 export async function importWidgetModule(rs: RpcSessionAtPos, pos: DocumentPosition, hash: string): Promise<any> {
     if (moduleCache.has(hash)) {
         // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        return moduleCache.get(hash)!
+        const [mod, _] = moduleCache.get(hash)!
+        return mod
     }
     const resp = await Widget_getWidgetSource(rs, pos, hash)
-    const mod = await dynamicallyLoadModule(hash, resp.sourcetext)
-    moduleCache.set(hash, mod)
+    let src = resp.sourcetext
+
+    /*
+     * Now we want to handle imports of other `@[widget_module]`s in `src`.
+     * At least two ways of doing this are possible:
+     * 1. Set a module resolution hook in `es-module-shims` to look through a global list of resolvers,
+     *    and register such a resolver here before loading a new module.
+     *    The resolver would add appropriate entries into the import map
+     *    before `src` is loaded and makes use of those entries.
+     *    However, resolution hooking and dynamic import maps are not standard features
+     *    so necessarily require `es-module-shims`;
+     *    they would not work with any current browser's ES module implementation.
+     *    Furthermore, this variant involves complex global state.
+     * 2. Before loading the module, parse its imports,
+     *    recursively import any widget modules,
+     *    and replace widget module imports with `blob:` URIs.
+     *    We do this as it is independent of `es-module-shims`.
+     *    A disadvantage is that this variant does not modify the global import map,
+     *    so any module that is not imported as a widget module (e.g. is imported from NPM)
+     *    cannot import widget modules.
+     */
+
+    await esmlexer.init
+    const [imports] = esmlexer.parse(src)
+    for (const i of imports) {
+        const HASH_URI_SCHEME = 'widget_module:hash,'
+        if (i.n?.startsWith(HASH_URI_SCHEME)) {
+            const h = i.n.substring(HASH_URI_SCHEME.length)
+            await importWidgetModule(rs, pos, h)
+            // `moduleCache.has(h)` is a postcondition of `importWidgetModule`
+            const [_, uri] = moduleCache.get(h)!
+            /* HACK: `es-module-lexer` doesn't provide indices of `import`s into the module source,
+             * so just replace the `import`ed name wholesale wherever it appears.
+             * It will break any module that relies on having that string literal intact,
+             * but such modules ought to be extremely unusual. */
+            src = src.replace(i.n, uri)
```

Comment:
I'm probably missing something, but https://www.npmjs.com/package/es-module-lexer documents the existence of `i.s` and `i.e`. Why are those not applicable here?

Reply:
Oh you are missing nothing, on the other hand I can't read. Fixed this now.

Reply:
If you mutate `src`, are the positions still correct when the second import with a widget import is being processed?

Reply:
Ooops! Thank you.